hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/AsyncProtobufLogWriter.java

```diff
@@ -59,7 +59,7 @@
 
   private final Class<? extends Channel> channelClass;
 
-  private AsyncFSOutput output;
+  private volatile AsyncFSOutput output;
```

Comment:
Do we really need to make it volatile? IIRC, it will be initialized in the init call and we will always call it when constructing a new ProtobufLogWriter. Make it volatile may cause developer confusing...

Reply:
The `output` would be set to null in `AsyncProtobufLogWriter.close` and would be used in `getSyncedLength` method, and seems there is no explicit constraint on which thread could invoke `AsyncProtobufLogWriter.close` , so here use `volatile`

Reply:
We will also use output in other methods where we do not test whether it is null, only test it in this method seems a bit confusing...

Reply:
Yes, you are right, I ignored other `write` methods also not test whether `output` is null. The problem here is it is meaningless to call  `AsyncProtobufLogWriter.getSyncedLength` after `AsyncProtobufLogWriter.close` , which indeed occured when  #1970  is applied to branch-2 previously caused by problematic state synchronization.

If we do not test whether output is null, we can just leave some comments here.

Reply:
Oh, then this is another problem? I think it is allowed to call getSyncedLength after closing? We should not throw IllegalStateException...

Reply:
Yes ,the  problem description  is :
https://issues.apache.org/jira/browse/HBASE-24625?focusedCommentId=17152610&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17152610

In my opinion,` AsyncProtobufLogWriter.getSyncedLength` is just meaningful when the WAL file is current writing, once 
`AsyncProtobufLogWriter` is closed(BTW, `AsyncProtobufLogWriter.output` is null),  upperlayer code seems  no need to call this method, if upperlayer code still call this method after closed, it may indicates some  synchronization error or other.

 If we allow to call `getSyncedLength` after closing, we should save the `syncedLength` when  `AsyncProtobufLogWriter.close` , or else `getSyncedLength` will throw `NullPointException` or just return 0, both are unexpected and may cause upperlayer code error.

Reply:
In practise, as we will call getSyncedLength in another thread, it is possible that we call getSyncedLength on a closed writer. Maybe we could introduce another field to save the final length of the file, and in getSyncedLength, we check whether we have this field set, if so we just return the value of this field, otherwise we will go to get output. WDYT?

Reply:
Yes, I agree with you, and would modify PR following your suggestion.

Reply:
@Apache9 , I modify PR to introduce `finalSyncedLength` to save the syncedLength when `AsyncProtobufLogWriter.close`.

But in `getSyncedLength` method,  I check whether output is  null  instead of  checking whether `finalSyncedLength` is set, because I think the statement "this.output = null;" in `AsyncProtobufLogWriter.close` is a  good sync point, if output is null, then `finalSyncedLength` must set, so we can return `finalSyncedLength`, else we return `output.getSyncedLength`.

If we use `finalSyncedLength` as a sync point, it is hard to decide whether the `output `is null or not.

Reply:
@Apache9 , please have a review, thanks.